src/core/ddsc/src/dds_loan.c

```diff
@@ -0,0 +1,242 @@
+    struct dds_reader const *const rd = (struct dds_reader *)e;
+    // only if SHM is enabled correctly (i.e. iox subscriber is initialized) and
+    // the type is fixed
+    ret = (rd->m_iox_sub != NULL);
+    break;
+  }
+  case DDS_KIND_WRITER: {
+    struct dds_writer const *const wr = (struct dds_writer *)e;
+    // only if SHM is enabled correctly (i.e. iox publisher is initialized) and
+    // the type is fixed
+    ret = (wr->m_iox_pub != NULL);
+    break;
+  }
+  default:
+    break;
+  }
+
+  dds_entity_unpin(e);
+#endif
+  (void)entity;
+  return ret;
+}
+
+bool dds_is_loan_available(const dds_entity_t entity) {
+  bool ret = false;
+#ifdef DDS_HAS_SHM
+  dds_entity *e;
+
+  if (DDS_RETCODE_OK != dds_entity_pin(entity, &e)) {
+    return ret;
+  }
+
+  switch (dds_entity_kind(e)) {
+  case DDS_KIND_READER: {
+    struct dds_reader const *const rd = (struct dds_reader *)e;
+    // only if SHM is enabled correctly (i.e. iox subscriber is initialized) and
+    // the type is fixed
+    ret = (rd->m_iox_sub != NULL) && (rd->m_topic->m_stype->fixed_size);
+    break;
+  }
+  case DDS_KIND_WRITER: {
+    struct dds_writer const *const wr = (struct dds_writer *)e;
+    // only if SHM is enabled correctly (i.e. iox publisher is initialized) and
+    // the type is fixed
+    ret = (wr->m_iox_pub != NULL) && (wr->m_topic->m_stype->fixed_size);
+    break;
+  }
+  default:
+    break;
+  }
+
+  dds_entity_unpin(e);
+#endif
+  (void)entity;
+  return ret;
+}
+
+bool is_loan_available(const dds_entity_t entity) {
+  return dds_is_loan_available(entity);
+}
+
+#ifdef DDS_HAS_SHM
+
+static void release_iox_chunk(dds_writer *wr, void *sample) {
+  iox_pub_release_chunk(wr->m_iox_pub, sample);
+}
+
+void register_pub_loan(dds_writer *wr, void *pub_loan) {
+  for (uint32_t i = 0; i < MAX_PUB_LOANS; ++i) {
+    if (!wr->m_iox_pub_loans[i]) {
+      wr->m_iox_pub_loans[i] = pub_loan;
+      return;
+    }
+  }
+  /* The loan pool should be big enough to store the maximum number of open
+   * IceOryx loans. So if IceOryx grants the loan, we should be able to store
+   * it.
+   */
+  assert(false);
+}
+
+bool deregister_pub_loan(dds_writer *wr, const void *pub_loan) {
+  for (uint32_t i = 0; i < MAX_PUB_LOANS; ++i) {
+    if (wr->m_iox_pub_loans[i] == pub_loan) {
+      wr->m_iox_pub_loans[i] = NULL;
+      return true;
+    }
+  }
+  return false;
+}
+
+static void *dds_writer_loan_chunk(dds_writer *wr, size_t size) {
+  void *chunk = shm_create_chunk(wr->m_iox_pub, size);
+  if (chunk) {
+    register_pub_loan(wr, chunk);
+    // NB: we set this since the user can use this chunk not only with write
+    // where we check whether it was loaned before.
+    // It is only possible to loan for fixed size types as of now.
+
+    // Unfortunate, since the chunk is not actually filled at this point.
+    // We should ensure that we cannot circumvent the write API
+    // (API redesign).
+    shm_set_data_state(chunk, IOX_CHUNK_CONTAINS_RAW_DATA);
```

Comment:
I think there is a good case for setting to `RAW_DATA` here: the interface is meant to be used for transferring the samples in the raw "in-memory" representation of the application, so it might as well be set accordingly. Alternatively, it could be set to `UNINITIALIZED` here and then overwritten with `RAW_DATA` in `dds_write`, but then one wonders: what's the point?

It all depends on how one interprets the "data state": does it encode only the representation, or does it also encode the validity of the contents? Both are fine; I have interpreted it as the former, and then setting it to `RAW_DATA` here is a sensible choice.

Reply:
The problem is that we get the chunk with `dds_loan_sample` and the user is supposed to fill it with a sample of a fixed size type. In `dds_write` we rely on the state to be correctly set. If it would ust be uninitialized here, the user would have to set the state and this is inconvenient and even more error-prone.

The only way to avoid this properly is having a function to fill the data to be called by the user. In C++ this would be a (variadic) emplace consctructor which also sets the state after the sample was intialized. I do not see this to be viable here, so the current state is the best alternative I came up with ...

Reply:
Since this function is static and only used in `dds_loan_sample()`, this should be fine